### PR TITLE
Refactoring internal test/wait

### PR DESCRIPTION
**Pull request by danghvu**
This is the third attempt to make this a clean PR (#2742, #2951):

* Make series of `MPIR_Wait/Test*` taking `MPIR_Request` as default implementation of MPID_Wait/Test*
* Assume the device will set `flag` (Test), `outcount` and `index/array_of_indices` (`*any/some`), MPI runtime only handle status and error checking.
* Make a new `MPIR_Request_is_active` since Test/Wait any/some only records "active" request, and there is no way to know without `MPIR_Request_complete` it.
* Make a new `MPIR_Request_wait_and_complete` to reduce duplicated code in collective when the non-blocking version is used as the implementation of blocking one.
* Make each commit self-contained.
* Co-authored-by @hajimefu (please help review some of the code in any/some cases which I tried to make the device do the work instead of duplicating work)

### CODE REVIEW

**Comment by hajimefu** [comment]
@danghvu Minor comments on authorship: the following changes were done by @michael-chuvelev, so could you include him as an author instead of me?
* Change MPIR_Test*_impl semantics 
* Introduce MPIR_Waitsome_impl

Do you plan to apply the batching idea to other array-based calls?

File: src/mpi/request/testall.c [hunk 1]
```diff
@@ -265,13 +159,126 @@ int MPI_Testall(int count, MPI_Request array_of_requests[], int *flag,
...
+                    status_ptr = (array_of_statuses != MPI_STATUSES_IGNORE) ? &array_of_statuses[i] : MPI_STATUS_IGNORE;
+                    if (status_ptr != MPI_STATUS_IGNORE) status_ptr->MPI_ERROR = rc;
+                    proc_failure = TRUE;
+                }
+            }
+        }
+    }
+
+    /* All requests have completed, or an error has occured, finalize or set errors. */
+    if (*flag == TRUE || mpi_errno == MPI_ERR_IN_STATUS)
```
Comment (hajimefu) [inline src/mpi/request/testall.c]: Since we are treating`flag` as a boolean I'd prefer `!*flag` or `*flag != FALSE` instead of `*flag == TRUE`.

**Comment by danghvu** [comment]
Sure, I will update to add him. About applying to other, I am not sure, I need to write a benchmark and this PR is already large enough.

The nbh collective test is failing, I may have missed a patch when resolving the rebase conflict, any idea?

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/waitall.c [hunk 2]
```diff
@@ -67,57 +67,25 @@ static inline int request_complete_fastpath(MPI_Request *request, MPIR_Request *
...
-    {
-        MPIR_CHKLMEM_MALLOC(request_ptrs, MPIR_Request **, count * sizeof(MPIR_Request *), mpi_errno, "request pointers", MPL_MEM_OBJECT);
-    }
 
     n_greqs = 0;
     n_completed = 0;
     for (i = 0; i < count; i++)
     {
-	if (array_of_requests[i] != MPI_REQUEST_NULL)
+        if (request_ptrs[i] != NULL)
```
Comment (hajimefu) [inline src/mpi/request/waitall.c]: nitpick: indent is off by one? (I wouldn't worry too much about the style as we are going to apply the code cleanup script soon anyway.)

File: src/mpi/request/waitany.c [hunk 3]
```diff
@@ -71,14 +134,11 @@ int MPI_Waitany(int count, MPI_Request array_of_requests[], int *indx,
...
-    MPID_Progress_state progress_state;
     int i;
-    int n_inactive;
     int active_flag;
-    int init_req_array;
-    int found_nonnull_req;
-    int last_disabled_anysource = -1;
+    int found_nonnull_req = FALSE;
     int mpi_errno = MPI_SUCCESS;
+    int off = 0;
```
Comment (hajimefu) [inline src/mpi/request/waitany.c]: Is this used?

File: src/include/mpir_request.h [hunk 4]
```diff
@@ -87,6 +87,14 @@ typedef struct MPIR_Grequest_class {
 
 #define MPIR_Request_is_complete(req_) (MPIR_cc_is_complete((req_)->cc_ptr))
 
+/* A persistent request and the handle to it are "inactive"
+ * if the request is not associated with any ongoing communication.
+ * A handle is "active" if it is neither null nor "inactive". */
+#define MPIR_Request_is_active(req_) ((req_) != NULL &&                 \
```
[the remark this answers is not in the document]
Comment (danghvu) [reply src/include/mpir_request.h]: Yes, the reason is only to be consistent with the `is_complete` just above it. I can change to static inline maybe with the MPL prefix to force inline.

File: src/mpi/request/waitany.c (hunk 3 above)
Comment (danghvu) [reply src/mpi/request/waitany.c]: Nope, left over from the old code, will remove -- thanks!

File: src/mpi/request/testall.c (hunk 1 above)
Comment (danghvu) [reply src/mpi/request/testall.c]: Yes I can change to `!= FALSE` -- I don't want to use just `*flag` because there is no guarantee that one will define `FALSE == 0` (just being paranoid here)

File: src/mpi/request/testall.c [hunk 5]
```diff
@@ -35,148 +35,34 @@ int MPI_Testall(int count, MPI_Request array_of_requests[], int *flag,
...
-                    }
-                }
+            if (!MPIR_Request_is_complete(request_ptrs[i])) {
+                goto fn_exit;
             }
         }
     }
-
-    *flag = (n_completed == count) ? TRUE : FALSE;
+    *flag = TRUE;
```
Comment (halimamer) [inline src/mpi/request/testall.c]: `flag=TRUE` implies all requested have completed. This commit breaks this semantic by dropping `n_completed` and unconditionally setting `flag`

Comment (danghvu) [reply src/mpi/request/testall.c]: Not sure if I understand this, if there is an uncompleted request It should have return and never arrive here.

Comment (halimamer) [reply src/mpi/request/testall.c]: right, never mind. I thought I was looking at testany/some

**Comment by halimamer** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/testall.c (hunk 5 above)
Comment (halimamer) [reply src/mpi/request/testall.c]: but `n_complete` is still unused. Can you compile locally with `--enable-strict` warnings of this type?

Comment (danghvu) [reply src/mpi/request/testall.c]: Yes, I'm cleaning those up. Thanks for the useful flag!

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
tests are now green, please help review.

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by pavanbalaji** [comment]
Thanks, @danghvu.

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/testsome.c [hunk 6]
```diff
@@ -31,11 +31,53 @@ int MPI_Testsome(int incount, MPI_Request array_of_requests[], int *outcount,
 #undef MPI_Testsome
 #define MPI_Testsome PMPI_Testsome
 
+#undef FUNCNAME
```
Comment (hajimefu) [inline src/mpi/request/testsome.c]: I'm not sure if I authored this patch. Perhaps only Michael. According to #2742 I did not.

File: src/mpi/request/waitall.c [hunk 7]
```diff
@@ -153,11 +153,12 @@ int MPIR_Waitall_impl(int count, MPIR_Request * request_ptrs[], MPI_Status array
   Setting statuses, freeing requests objects, etc.
 */
 #undef FUNCNAME
-#define FUNCNAME MPIR_Waitall_post
+#define FUNCNAME do_waitall_post
```
Comment (hajimefu) [inline src/mpi/request/waitall.c]: Maybe we should squash this into the waitall patch?

File: src/mpi/request/waitall.c [hunk 8]
```diff
@@ -25,13 +25,6 @@ int MPI_Waitall(int count, MPI_Request array_of_requests[], MPI_Status array_of_
 #endif
 /* -- End Profiling Symbol Block */
 
-/* Define MPICH_MPI_FROM_PMPI if weak symbols are not supported to build
```
Comment (hajimefu) [inline src/mpi/request/waitall.c]: In the commit message, can you add a little bit more explanation -- why this was needed, where the cache miss happened, and how this patch would solve it?

**Comment by hajimefu** [comment]
Went through the code again.

Overall the code looks okay. Left some minor comments. I assume you applied the latest code cleanup/style checker? At least I did not see any issue when running it locally.

**Comment by danghvu** [comment]
Thanks @hajimefu I ran the style check/cleanup once. I will do it again after the other patch (request management) goes in and this one is rebased on top of it.

File: src/mpi/request/testany.c [hunk 9]
```diff
@@ -30,10 +30,52 @@ int MPI_Testany(int count, MPI_Request array_of_requests[], int *indx, int *flag
...
+
 #endif
 
 #undef FUNCNAME
 #define FUNCNAME MPI_Testany
+#undef FCNAME
+#define FCNAME MPL_QUOTE(FUNCNAME)
 
 /*@
     MPI_Testany - Tests for completion of any previdously initiated
```
Comment (hajimefu) [inline src/mpi/request/testany.c]: typo: *previously

Fix to this should be in a separate commit.

File: src/mpi/request/testsome.c (hunk 6 above)
Comment (danghvu) [reply src/mpi/request/testsome.c]: fixed

File: src/mpi/request/waitall.c (hunk 8 above)
Comment (danghvu) [reply src/mpi/request/waitall.c]: added

File: src/mpi/request/testany.c (hunk 9 above)
Comment (danghvu) [reply src/mpi/request/testany.c]: I think this needs to be in a separate PR (maybe check typo for everything), this one is already too large.

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/waitall.c (hunk 7 above)
Comment (danghvu) [reply src/mpi/request/waitall.c]: done.

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by danghvu** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

File: src/mpi/request/test.c [hunk 10]
```diff
@@ -64,23 +52,9 @@ int MPIR_Test_impl(MPI_Request * request, int *flag, MPI_Status * status)
...
 
     if (MPIR_Request_is_complete(request_ptr)) {
-        mpi_errno = MPIR_Request_completion_processing(request_ptr, status, &active_flag);
-        if (!MPIR_Request_is_persistent(request_ptr)) {
-            MPIR_Request_free(request_ptr);
-            *request = MPI_REQUEST_NULL;
-        }
-        if (mpi_errno)
-            MPIR_ERR_POP(mpi_errno);
         *flag = TRUE;
```
Comment (pavanbalaji) [inline src/mpi/request/test.c]: What's the point of this flag?  All it's doing is checking if `MPIR_Request_is_complete` is true.  Whoever is calling this function can just check that instead of checking the flag.

File: src/mpi/request/test.c [hunk 11]
```diff
@@ -161,12 +136,36 @@ int MPI_Test(MPI_Request * request, int *flag, MPI_Status * status)
     }
 #endif /* HAVE_ERROR_CHECKING */
 
+    /* If this is a null request handle, then return an empty status */
+    if (*request == MPI_REQUEST_NULL) {
+        MPIR_Status_set_empty(status);
+        *flag = TRUE;
+        goto fn_exit;
+    }
```
Comment (pavanbalaji) [inline src/mpi/request/test.c]: Moving this outside the `MPIR_Test_impl` function means that every function that is calling it would need to explicitly check for this case.  In this patch, for example, `MPIR_Wait_impl` is not checking for this.

File: src/mpi/request/test.c (hunk 10 above)
Comment (danghvu) [reply src/mpi/request/test.c]: The flag is the user's provided flag passed from upper layer, either this will set this, or the upper layer has to set it. I guess it boils down to what we want the "device" layer to do.

File: src/mpi/request/test.c (hunk 11 above)
Comment (danghvu) [reply src/mpi/request/test.c]: `MPIR_Wait_impl` is changed in a different patch, the original code isn't consistent, but the final code I believe both were checking this outside.
https://github.com/pmodels/mpich/pull/2951/files#diff-2023d8a87eecda777a2ff89cc5f57272L163

File: src/mpi/request/test.c (hunk 10 above)
Comment (pavanbalaji) [reply src/mpi/request/test.c]: It makes sense for the MPI interface, since the user has no other way of checking for completion.  But it doesn't make sense for the MPID interface (or in this case, internal to the MPIR layer), since the `MPIR_Request_is_complete` function exists to do that.

File: src/mpi/request/test.c (hunk 11 above)
Comment (pavanbalaji) [reply src/mpi/request/test.c]: What do you mean by the original code not being consistent?  Your first patch is introducing a bug that didn't exist in the original code.  Each patch should be consistent.  Otherwise, there won't be any point of separating those patches.

File: src/mpi/request/test.c (hunk 10 above)
Comment (danghvu) [reply src/mpi/request/test.c]: What do you suggest for the device layer then ?

Comment (pavanbalaji) [reply src/mpi/request/test.c]: You refer to the "device layer", which is a bit confusing.  In the patch I'm talking about, both functions are in the MPI layer.  And one function is announcing the completion of the request in two ways: through the request itself and through the flag.  I'm just saying that it's redundant.  One of them is sufficient.

File: src/mpi/request/test.c (hunk 11 above)
Comment (danghvu) [reply src/mpi/request/test.c]: There is no bug, since MPI wait have always checked for this expept for the test check in test_impl and wait check outside originally

Comment (danghvu) [reply src/mpi/request/test.c]: There is no bug, since MPI wait have always checked for this expept for the test check in test_impl and wait check outside originally

Comment (pavanbalaji) [reply src/mpi/request/test.c]: Oh, yes, you are right.  The check was redundant in `MPI_Wait`, so removing it from `MPIR_Test_impl` doesn't get rid of that functionality.  I withdraw this comment.

**Comment by danghvu** [comment]
We are abandoning this for pr #2980, if you are interested please take a look at the other PR.
